### ISSUE TITLE
作者您好，请问误判情况该怎么处理

### ISSUE THREAD

**Issue by stephencurry2** [qcf-568/DocTamper#81]
![image](https://github.com/user-attachments/assets/ccdb3c72-c198-415d-b1d7-c868a7911631)
![image](https://github.com/user-attachments/assets/02e1866f-fa54-4d60-a002-4ccaaa739186)
![image](https://github.com/user-attachments/assets/334b81bc-edfa-4f54-bffa-d33dc46e5153)
作者您好，我是使用了您的模型，测试了我的图片，发现有许多误判，请问这是怎么回事呢

**Comment by qcf-568** [comment]
您好，请问您的推理代码是怎样的

**Comment by stephencurry2** [comment]
您好 我的推理代码这样的：
def crop_img(img, jpg_dct, crop_size=512, mask=None):
    if mask is None:
        use_mask = False
    else:
        use_mask = True
        crop_masks = []

    h, w, c = img.shape
    h_grids = h // crop_size
    w_grids = w // crop_size

    crop_imgs = []
    crop_jpe_dcts = []

    for h_idx in range(h_grids):
        for w_idx in range(w_grids):
            x1 = w_idx * crop_size
            x2 = x1 + crop_size
            y1 = h_idx * crop_size
            y2 = y1 + crop_size
            crop_img = img[y1:y2, x1:x2, :]
            crop_imgs.append(crop_img)
            crop_jpe_dct = jpg_dct[y1:y2, x1:x2]
            crop_jpe_dcts.append(crop_jpe_dct)
            if use_mask:
                if mask[y1:y2, x1:x2].max() != 0:
                    crop_masks.append(1)
                else:
                    crop_masks.append(0)

    if w % crop_size != 0:
        for h_idx in range(h_grids):
            y1 = h_idx * crop_size
            y2 = y1 + crop_size
            crop_imgs.append(img[y1:y2, w - 512:w, :])
            crop_jpe_dcts.append(jpg_dct[y1:y2, w - 512:w])
            if use_mask:
                if mask[y1:y2, w - 512:w].max() != 0:
                    crop_masks.append(1)
                else:
                    crop_masks.append(0)

    if h % crop_size != 0:
        for w_idx in range(w_grids):
            x1 = w_idx * crop_size
            x2 = x1 + crop_size
            crop_imgs.append(img[h - 512:h, x1:x2, :])
            crop_jpe_dcts.append(jpg_dct[h - 512:h, x1:x2])
            if use_mask:
                if mask[h - 512:h, x1:x2].max() != 0:
                    crop_masks.append(1)
                else:
                    crop_masks.append(0)

    if w % crop_size != 0 and h % crop_size != 0:
        crop_imgs.append(img[h - 512:h, w - 512:w, :])
        crop_jpe_dcts.append(jpg_dct[h - 512:h, w - 512:w])
        if use_mask:
            if mask[h - 512:h, w - 512:w].max() != 0:
                crop_masks.append(1)
            else:
                crop_masks.append(0)

    if use_mask:
        return crop_imgs, crop_jpe_dcts, h_grids, w_grids, crop_masks
    else:
        return crop_imgs, crop_jpe_dcts, h_grids, w_grids, None


def combine_img(imgs, h_grids, w_grids, img_h, img_w, crop_size=512):
    i = 0
    re_img = np.zeros((img_h, img_w))
    for h_idx in range(h_grids):
        for w_idx in range(w_grids):
            x1 = w_idx * crop_size
            x2 = x1 + crop_size
            y1 = h_idx * crop_size
            y2 = y1 + crop_size
            re_img[y1:y2, x1:x2] = imgs[i]
            i += 1

    if w_grids * crop_size < img_w:
        for h_idx in range(h_grids):
            y1 = h_idx * crop_size
            y2 = y1 + crop_size
            re_img[y1:y2, img_w - 512:img_w] = imgs[i]
            i += 1

    if h_grids * crop_size < img_h:
        for w_idx in range(w_grids):
            x1 = w_idx * crop_size
            x2 = x1 + crop_size
            re_img[img_h - 512:img_h, x1:x2] = imgs[i]
            i += 1

    if w_grids * crop_size < img_w and h_grids * crop_size < img_h:
        re_img[img_h - 512:img_h, img_w - 512:img_w] = imgs[i]

    return re_img

def pdf_to_images(pdf_path, output_folder):
    images = convert_from_path(pdf_path)
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)
    for idx, img in enumerate(images):
        img_path = os.path.join(output_folder, f"page_{idx + 1}.jpg")
        img.save(img_path, "JPEG")
    return output_folder

def get_page_number_from_filename(filename):

    if filename.startswith('page_') and filename.endswith('.jpg'):
        page_number_str = filename[len('page_'):-len('.jpg')]
        return int(page_number_str)
    return None



def detect_ps_in_pdf(pdf_path, crop_size=512, threshold=0.5):
   

    device = torch.device("cuda")
    #print(f"Using device: {device}")
    model = seg_dtd("", 2).to(device)
    model = nn.DataParallel(model)
    # following executes the same mentioned above
    for name, module in model.named_modules():
        if isinstance(module, nn.GELU):
            exec('model.' + torchmodify(name) + '=nn.GELU()')
    loader = torch.load(args.weight, map_location='cpu')['state_dict']
    model.load_state_dict(loader)

    new_qtb = np.array(
        [[2, 1, 1, 2, 2, 4, 5, 6], [1, 1, 1, 2, 3, 6, 6, 6], [1, 1, 2, 2, 4, 6, 7, 6], [1, 2, 2, 3, 5, 9, 8, 6],
         [2, 2, 4, 6, 7, 11, 10, 8], [2, 4, 6, 6, 8, 10, 11, 9], [5, 6, 8, 9, 10, 12, 12, 10],
         [7, 9, 10, 10, 11, 10, 10, 10]], dtype=np.int32).reshape(64, ).tolist()

    data_path = args.input
    result_path = args.output

    if not os.path.exists(result_path):
        os.mkdir(result_path)


    totsr = ToTensorV2()
    toctsr = torchvision.transforms.Compose([
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize(mean=(0.485, 0.455, 0.406), std=(0.229, 0.224, 0.225))
    ])

    result_path = args.output
    if not os.path.exists(result_path):
        os.makedirs(result_path)

    #pdf_path = args.pdf_path
    output_folder = args.output_folder
    images_folder = pdf_to_images(pdf_path, output_folder)

    model.eval()
    crop_masks_alls = []
    pred_lists_alls = []
    data_path = args.input
    result_path = args.output
    ps_detected_pages = []
    page_ps_probabilities = []
    page_heatmaps = []
    ps_traces = []

    for path in tqdm(os.listdir(output_folder)):
        if str(path).endswith(("jpg", 'jpeg', 'JPG', 'JPEG')):
            page_idx = get_page_number_from_filename(path)
            # print(page_idx)

            img_path = os.path.join(data_path, path)
            # print(img_path)
            imgs_ori = cv2.imread(img_path)
            yuantu = imgs_ori
            h, w, c = imgs_ori.shape
            H1, W1, C1 = imgs_ori.shape
            jpg_dct = jpegio.read(img_path)
            gt_mask = cv2.imread('test_masks/' + path[:-4] + '.png', 0)
            dct_ori = jpg_dct.coef_arrays[0].copy()
            use_qtb2 = jpg_dct.quant_tables[0].copy()

            if min(h, w) < 512:
                H, W = imgs_ori.shape[:2]
                if H < 512:
                    dh = (512 - H)
                else:
                    dh = 0
                if W < 512:
                    dw = (512 - W)
                else:
                    dw = 0

                imgs_ori = np.pad(imgs_ori, ((0, dh), (0, dw), (0, 0)), 'constant', constant_values=255)
                with tempfile.NamedTemporaryFile(delete=True) as tmp:
                    imgs_ori = Image.fromarray(imgs_ori).convert("L")
                    imgs_ori.save(tmp, "JPEG", qtables={0: new_qtb})
                    jpg = jpegio.read(tmp.name)
                    dct_ori = jpg.coef_arrays[0].copy()
                    imgs_ori = np.array(imgs_ori.convert('RGB'))
                    use_qtb2 = jpg.quant_tables[0].copy()
                h, w, c = imgs_ori.shape

            if h % 8 == 0 and w % 8 == 0:
                imgs_d = imgs_ori
                dct_d = dct_ori
            else:
                imgs_d = imgs_ori[0:(h // 8) * 8, 0:(w // 8) * 8, :].copy()
                dct_d = dct_ori[0:(h // 8) * 8, 0:(w // 8) * 8].copy()

            qs = torch.LongTensor(use_qtb2)
            img_h, img_w, _ = imgs_d.shape
            crop_imgs, crop_jpe_dcts, h_grids, w_grids, _ = crop_img(imgs_d, dct_d, crop_size=512, mask=gt_mask)
            img_list = []
            has_ps_traces = False

           
            for idx, crop in enumerate(crop_imgs):
                crop = Image.fromarray(cv2.cvtColor(crop, cv2.COLOR_BGR2RGB))
                data = toctsr(crop)
                dct = torch.LongTensor(crop_jpe_dcts[idx])

                data, dct, qs = data.unsqueeze(0).to(device), dct.unsqueeze(0).to(device), qs.unsqueeze(0).to(device)
                dct = torch.abs(dct).clamp(0, 20)
                B, C, H, W = data.shape
                qs = qs.reshape(B, 1, 8, 8)
                
                with torch.no_grad():
                    if data.size()[-2:] == torch.Size((512, 512)) and dct.size()[-2:] == torch.Size(
                            (512, 512)) and qs.size()[-2:] == torch.Size((8, 8)):
                        pred = model(data, dct, qs)
                        pred = torch.nn.functional.softmax(pred, 1)[:, 1].cpu()
                        # print(pred.cpu().numpy())
                        pred_np = pred.cpu().numpy()
                        average_probability = non_zero_probabilities.mean()
                        img_list.append((pred_np * 255).astype(np.uint8))
            if  average_probability > 0:
                has_ps_traces = True
            if has_ps_traces:
                # print("", average_probability)
                ci = combine_img(img_list, h_grids, w_grids, img_h, img_w, crop_size=512)
                padding = (0, 0, w - img_w, h - img_h)
                ci = cv2.copyMakeBorder(ci, padding[1], padding[3], padding[0], padding[2], cv2.BORDER_CONSTANT,
                                        value=[0, 0, 0])
                ci = ci[0:H1, 0:W1]
                # cv2.imwrite(os.path.join(result_path, path.split(".jpg")[0]+".png"), ci)
                ci = np.uint8(ci)
                heatmap = cv2.applyColorMap(ci, cv2.COLORMAP_JET)
                overlay = cv2.addWeighted(yuantu, 0.7, heatmap, 0.3, 0)
                cv2.imwrite(os.path.join(result_path, path.split(".jpg")[0] + ".png"), overlay)
                save_path = os.path.join(result_path, path.split(".jpg")[0] + ".png")
                # cv2.imwrite(os.path.join(result_path, path.split(".jpg")[0] + ".png"), overlay)
                ps_trace = {
                    "page_idx": page_idx,
                    "probability": average_probability,
                    "heatmap_path": save_path
                }
                ps_traces.append(ps_trace)
                
    if (len(ps_traces)):
        return ps_traces
    else:
        print("No traces of PS were detected")
        return ps_traces

**Comment by qcf-568** [comment]
您好，经初步查看，认为问题所在是:
所提供的数据及模型暂不擅长对PDF文档进行处理。DocTamper数据集仅包含拍照文档，因此训练的模型无法处理PDF文档；另一方面，DTD模型本身设计是针对拍照文档场景，不擅长处理PDF文档，加之没有经过PDF文档数据训练，因此对PDF场景表现出误检多的情况。

所需要解决纯PDF场景篡改检测问题，建议使用纯视觉域模型，如[纯视觉域检测模型](https://github.com/qcf-568/Det_Baseline)或[纯视觉域分割模型](https://github.com/qcf-568/MIML)，若要同时解决拍照及PDF场景，则需要[Omni-IML](https://arxiv.org/pdf/2411.14823)。上述模型均需要在包含待解决的PDF场景训练数据中重新训练。

**Comment by stephencurry2** [comment]
好的 谢谢您 还有就是我想问一下如果我想完成检测pdf篡改检测任务，要自己对pdf里的图片进行篡改并制作相应的掩码来制作训练数据集然后训练吗

**Comment by qcf-568** [comment]
是的，要有把PDF转化为PNG及JPG，制作相应的PDF篡改训练数据微调才行。
